src/EditorFeatures/Core/Diagnostics/DiagnosticsClassificationTaggerProvider.cs

```diff
@@ -62,18 +61,61 @@ protected internal override bool IsEnabled
             => !_editorOptionsService.Factory.GlobalOptions.GetOptionValue(DefaultTextViewHostOptions.IsInContrastModeId);
 
         protected internal override bool IncludeDiagnostic(DiagnosticData data)
-            => data.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary);
+        {
+            // Fading diagnostic might have an 'Unnecessary' custom tag or
+            // unnecessary additional location indices.
+            if (!data.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary)
+                && !TryGetUnnecessaryLocationIndices(data, out _))
+            {
+                return false;
+            }
+
+            // Do not fade if user has disabled the fading option corresponding to this diagnostic.
+            PerLanguageOption2<bool> fadingOption;
+            switch (data.Id)
+            {
+                case IDEDiagnosticIds.RemoveUnnecessaryImportsDiagnosticId:
+                    fadingOption = IdeAnalyzerOptionsStorage.FadeOutUnusedImports;
+                    break;
+
+                case IDEDiagnosticIds.RemoveUnreachableCodeDiagnosticId:
+                    fadingOption = IdeAnalyzerOptionsStorage.FadeOutUnreachableCode;
+                    break;
+
+                case IDEDiagnosticIds.UseObjectInitializerDiagnosticId:
+                    fadingOption = IdeAnalyzerOptionsStorage.FadeOutComplexObjectInitialization;
+                    break;
+
+                case IDEDiagnosticIds.UseCollectionInitializerDiagnosticId:
+                    fadingOption = IdeAnalyzerOptionsStorage.FadeOutComplexCollectionInitialization;
+                    break;
+
+                default:
+                    return true;
```

Comment:
i also don't like that htis returns true without doing hte lang/fadign check below.

Reply:
The `true` here as I understand are for fading that's not controlled by any options. For example, the VB analyzer that removes `ByVal` does fading without an option. There is no language/fading check to be done.

Reply:
> we should doc that this is in intentional.

definitely, I can add more comments here.

> it woudl be good if there was an Contract.Fail here that triggers so that if someone adds something new, they then are forced to think about updating this code.

I don't think this would be possible as the custom tag is exposed publicly and even third party analyzers can report diagnostics with an Unnecessary tag. Additionally, I don't know if it would be common to support such fading options in future, so enumerating all IDE diagnostic IDs here might just become a pain point.